lib/eco/autolboxdetector.py

```diff
@@ -42,24 +42,98 @@ def pop_lookahead(self, node):
             node = node.parent
         return node.right_sibling()
 
-    def find_terminal(self, node):
+    def find_terminal(self, node, version=None):
         startnode = node
         while type(node) is not BOS:
-            if node.children:
-                node = node.children[-1]
+            if node.get_attr("children", version):
+                node = node.get_attr("children", version)[-1]
             elif type(node.symbol) is Terminal:
                 break
             else:
                 if node.new:
                     return
-                while not node.left_sibling():
-                    node = node.parent
+                while not node.get_attr("left", version):
+                    node = node.get_attr("parent", version)
                     if node is startnode:
                         return None
-                node = node.left_sibling()
+                node = node.get_attr("left", version)
         return node.next_term
 
     def detect_lbox(self, errornode):
+        # Try heuristic #2 first; if it fails to find an automatic language
+        # box, try heuristic #1
+        valid = self.detect_lbox_h2(errornode)
+        if not valid:
+            valid = self.detect_lbox_h1(errornode)
+
+        if errornode.autobox is False:
+            # XXX Currently, we don't suggest any language boxes for an error,
+            # if the user had to revert an automatically inserted box. However,
+            # we might want to consider showing suggestions if the suggested box
+            # differs from the one the user reverted
+            return False # don't use this node for autoboxes anymore
+        if valid:
```

Comment:
What are the possible values of `valid` here? `True/False/None`...?

Reply:
Valid is a list which contains all candidates found by the heuristics. Should I change this to `if len(valid) > 0`?

Reply:
Ping.

Reply:
Yes, definitely. [Boolean checks in Python are a disaster: the empty list comparing equal to `False` is a terrible decision!]